### PR TITLE
Checkpoint

### PR DESCRIPTION
**Pull request by vreuter**
Units are looking good, but I've not yet _fully_ integrated this into a true pipeline, e.g. `chipseq.py`,. 
Wanted to go ahead & open for review, though, since the tests are behaving.
The elements to focus on, IMO, are `stage.Stage`, `pipeline.Pipeline`, and updates to `PipelineManager` (now in `manager.py`)
Close https://github.com/epigen/pypiper/issues/56

The value-added updates in this PR are:
- conceptual grouping into stages, which are defined as functions
- stages can process arbitrary functions (PipelineManager only processes shell commands)
- functions are checkpointed, and checkpoint targets are automatic (so stages are checkpointed automatically; all the user has to do is group functions into stages and pass this to the PIpeline)
- arbitrary start/stop of the pipeline via command-line options (this is the original motivation for the checkpoint concept)

### CODE REVIEW

**Comment by vreuter** [comment]
This is an implementation for https://github.com/epigen/pypiper/issues/56

**Comment by nsheff** [comment]
this version gives an erroneous message. Why is it generating recover files when target exists? there should be no lock or recovery files generated.

Here's an example from pepsra:

Expected output:
```
Processing 1 of 4

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030718.bam`
Processing 2 of 4

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030719.bam`
Processing 3 of 4

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030720.bam`
Processing 4 of 4

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030721.bam`
```

Actual output:
```
Processing 1 of 4
Generating recovery file from lock file: '/sfs/lustre/allocations/shefflab/data/sra/lock.__sfs__lustre__allocations__shefflab__data__sra_bam__SRR1030718.bam'

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030718.bam`
Processing 2 of 4
Generating recovery file from lock file: '/sfs/lustre/allocations/shefflab/data/sra/lock.__sfs__lustre__allocations__shefflab__data__sra_bam__SRR1030719.bam'

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030719.bam`
Processing 3 of 4
Generating recovery file from lock file: '/sfs/lustre/allocations/shefflab/data/sra/lock.__sfs__lustre__allocations__shefflab__data__sra_bam__SRR1030720.bam'

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030720.bam`
Processing 4 of 4
Generating recovery file from lock file: '/sfs/lustre/allocations/shefflab/data/sra/lock.__sfs__lustre__allocations__shefflab__data__sra_bam__SRR1030721.bam'

Target exists: `/sfs/lustre/allocations/shefflab/data/sra_bam/SRR1030721.bam`
```

**Comment by vreuter** [comment]
Sorry, good catch. The message is misleading; it's not actually creating the files (at least I think it's not & definitely shouldn't be); the message is being provided at time of derivation of that filepath.

**Comment by nsheff** [comment]
ok that fixed it.

**Comment by nsheff** [comment]
this is adding another error message:
```
Could not remove flag file: '/sfs/lustre/allocations/shefflab/data/sra/sra_convert_initializing.flag'
```

which was not previously showing up.

**Comment by nsheff** [comment]
It also adds another redundant error message:
```
Can't find custom config file: /sfs/nfs/blue/ns5bc/code/pepsra/sra_convert/sra_convert.yaml
No config file
```

previously it didn't say "No config file" as well.

**Comment by vreuter** [comment]
The config file message redundancy and flag removal message for initialization should be gone with the most recent commit.

File: pypiper/ngstk.py [hunk 1]
```diff
@@ -619,60 +771,134 @@ def count_fail_reads(self, file_name, paired_end):
...
+        if not (self.is_sam_or_bam(file_name) or self.is_fastq(file_name)):
+            # TODO: make this an exception and force caller to handle that
+            # rather than relying on knowledge of possibility of negative value.
+            return -1
+
+        if self.is_sam_or_bam(file_name):
+            param_text = "-c" if ext == ".bam" else "-c -S"
+            return self.samtools_view(file_name, param=param_text)
+        else:
+            count_lines = self.count_lines_zip \
```
Comment (nsheff) [inline pypiper/ngstk.py]: shouldn't this be `count_lines(file_name)` and `count_lines_zip(file_name)`?

**Comment by nsheff** [comment]
Counting raw fastq reads is broken:

```
> `Raw_reads`	-1	ATACseq	_RES_
> `Fastq_reads`	19486778	ATACseq	_RES_
Skipping undefined pypiper argument: 'sample_name'
Traceback (most recent call last):
  File "/home/ns5bc/code/ATACseq/pipelines/ATACseq.py", line 655, in <module>
    sys.exit(main())
  File "/home/ns5bc/code/ATACseq/pipelines/ATACseq.py", line 303, in main
    follow=ngstk.check_fastq(local_input_files, unaligned_fastq, args.paired_end))
  File "/sfs/nfs/blue/ns5bc/.local/lib/python2.7/site-packages/pypiper/manager.py", line 686, in run
    call_follow()
  File "/sfs/nfs/blue/ns5bc/.local/lib/python2.7/site-packages/pypiper/manager.py", line 619, in call_follow
    follow()
  File "/sfs/nfs/blue/ns5bc/.local/lib/python2.7/site-packages/pypiper/ngstk.py", line 507, in temp_func
    raise Exception("Fastq conversion error? Number of reads "
Exception: Fastq conversion error? Number of reads doesn't match unaligned bam
```
I flagged where I may have found the error but too much has changed for me to really diagnose...

This example case is specifically with gzipped fastq input.

**Comment by nsheff** [comment]
Closer but still not quite there;
> `Raw_reads`	1645650	
> `Fastq_reads`	47803236

**Comment by nsheff** [comment]
Ok, that should do it. it's working now.

**Comment by nsheff** [comment]
can you add a readme for devs on how we should run tests now? my old way of testing is no longer functional with this PR, and when I run 'pytest' it does not show the output as it used to -- it looks like it's still running the old tests, how can we make it show the output?

**Comment by nsheff** [comment]
can you explain the rationale for having both a Pipeline and a PipelineManager class? What is the difference? Why not just have 1 class?

I am not sure I would name this new class "pipeline", as to me, that is what the python script itself is, so there a conceptual disconnect at least in the way I think of things.

**Comment by vreuter** [comment]
Mostly separation of concerns. Generally, fewer and larger classes impedes code comprehensibility.

**Comment by vreuter** [comment]
Ok, I don’t have a strong name preference. Yes, it essentially represents a program/script. Like in the new chipseq version, there’s a main method that calls into the arguments parser but then is primarily responsible for building the pipeline instance and allowing it to handle execution

**Comment by nsheff** [comment]
both `Pipeline` and `PipelineManager` are user-facing, right? To me this adds confusion, I wouldn't understand why there are 2 classes, and what the difference is. I would favor a single user-facing class -- it doesn't have to be a huge class, I'm not opposed to subdividing the classes, but maybe we can do that internally and then present only single interface to the user?

the checkpoint is really just adding a single new function to pypiper, it doesn't seem to necessitate presenting the user with two classes

**Comment by nsheff** [comment]
Just to keep a note here for later: I think the cleaning is now too verbose ("1 items to clean: 0 items to clean: 1 items to clean:" etc)

**Comment by vreuter** [comment]
Agreed!

**Comment by nsheff** [comment]
Is this ready to be merged or are you still working on it?

**Comment by vreuter** [comment]
I'd like to write up a couple more tests, but I feel pretty good about it. If you have something small that you'd like to test with however you envision using the checkpointed timestamping and start/stop, that'd be a huge plus but I think not immediately necessary since this is just to `dev`.

**Comment by nsheff** [comment]
if you use `stop-after`, it calls `fail_pipeline` on every subsequent command call, and keeps changing status flags back and forth. it's doing weird stuff for me.

**Comment by nsheff** [comment]
I think I'm understanding better now; I think it needs to raise an exception to cause the pipeline to quit, so it doesn't keep progressing (even though it doesn't run subsequent commands, it should just stop and not even *attempt* to run them).

**Comment by vreuter** [comment]
Units and some in-person tests look good, merging with permission from @nsheff

**Pull request merged**